

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>GREEN-VARAN tool usage &mdash; GREEN-DB and GREEN-VARAN  documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="SV_annotation tool usage" href="SV_annotation_usage.html" />
    <link rel="prev" title="GREEN-VARAN tool set" href="GREEN_VARAN.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> GREEN-DB and GREEN-VARAN
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="GREEN_DB.html">The GREEN-DB</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="GREEN_VARAN.html">GREEN-VARAN tool set</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="GREEN_VARAN.html#installation">Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="GREEN_VARAN.html#requirements">Requirements</a></li>
<li class="toctree-l2"><a class="reference internal" href="GREEN_VARAN.html#singularity">Singularity</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="GREEN_VARAN.html#single-tools-usage">Single tools usage</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">GREEN-VARAN tool usage</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#basic-usage">Basic usage</a></li>
<li class="toctree-l4"><a class="reference internal" href="#prioritization">Prioritization</a></li>
<li class="toctree-l4"><a class="reference internal" href="#gene-mode">Gene mode</a></li>
<li class="toctree-l4"><a class="reference internal" href="#activate-additional-annotations">Activate additional annotations</a></li>
<li class="toctree-l4"><a class="reference internal" href="#fields-added-to-info">Fields added to INFO</a></li>
<li class="toctree-l4"><a class="reference internal" href="#arguments-list">Arguments list</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="SV_annotation_usage.html">SV_annotation tool usage</a></li>
<li class="toctree-l3"><a class="reference internal" href="GREEN-DB_query_usage.html">GREEN-DB_query tool usage</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="Download.html">Download resources</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">GREEN-DB and GREEN-VARAN</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="GREEN_VARAN.html">GREEN-VARAN tool set</a> &raquo;</li>
        
      <li>GREEN-VARAN tool usage</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/GREEN-VARAN_usage.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="green-varan-tool-usage">
<h1>GREEN-VARAN tool usage<a class="headerlink" href="#green-varan-tool-usage" title="Permalink to this headline">¶</a></h1>
<p>GREEN-VARAN.py performs annotation of small variants VCF.
It provides also abiliy to filter for genes of interest, select variants on genes already affected by coding variants
and a prioritization function for regulatory variants</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>GREEN-VARAN.py  <span class="o">[</span>-h<span class="o">]</span> -i VCF -o OUTPUT -b <span class="o">{</span>GRCh37,GRCh38<span class="o">}</span> -m
                <span class="o">{</span>annotate,filter_regdb,filter_any<span class="o">}</span> <span class="o">[</span>-p<span class="o">]</span>
                <span class="o">[</span>--vcfanno VCFANNO<span class="o">]</span> <span class="o">[</span>--bed_dir BED_DIR<span class="o">]</span>
                <span class="o">[</span>--AF_file AF_FILE<span class="o">]</span> <span class="o">[</span>--scores_dir SCORES_DIR<span class="o">]</span>
                <span class="o">[</span>-g <span class="o">{</span>annotate,filter<span class="o">}]</span> <span class="o">[</span>-t <span class="o">{</span>controlled,closest,both<span class="o">}]</span>
                <span class="o">[</span>-l GENES_LIST<span class="o">]</span> <span class="o">[</span>--impact <span class="o">{</span>HIGH,MODERATE,LOW,MODIFIER<span class="o">}]</span>
                <span class="o">[</span>--allelefreq <span class="o">{</span>global,afr,amr,eas,fin,nfe,sas,oth<span class="o">}]</span>
                <span class="o">[</span>-s <span class="o">{</span>ReMM,FIRE,LinSight,ExPECTO,NCBoost,DANN,CADD<span class="o">}]</span> <span class="o">[</span>-a<span class="o">]</span>
                <span class="o">[</span>--separate_fields<span class="o">]</span> <span class="o">[</span>--logfile LOGFILE<span class="o">]</span>
                <span class="o">[</span>--threads THREADS<span class="o">]</span> <span class="o">[</span>-w<span class="o">]</span>
</pre></div>
</div>
<div class="section" id="basic-usage">
<h2>Basic usage<a class="headerlink" href="#basic-usage" title="Permalink to this headline">¶</a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>GREEN-VARAN.py  <span class="o">[</span>-h<span class="o">]</span> -i VCF -o OUTPUT -b <span class="o">{</span>GRCh37,GRCh38<span class="o">}</span> -m
                <span class="o">{</span>annotate,filter_regdb,filter_any<span class="o">}</span>
</pre></div>
</div>
<p>The basic mode of run is to annotate variants with information from the GREEN-DB
and then output all variants or just annotated ones based on the selected running mode.</p>
<p>The running mode (-m, –mode) can be set to:</p>
<ul class="simple">
<li><dl class="simple">
<dt>annotate</dt><dd><p>In this mode the tool will perform a standard annotation.
It will annotate variants with required information and
then output all variants form the original VCF</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>filter_regdb</dt><dd><p>In this mode the tool will annotate variants with required information
but only variants that overlap one of the GREEN-DB regions will be outputted</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>filter_any</dt><dd><p>This mode acts similar to filter_regdb, but a variant is present in the output
if it overlaps a GREEN-DB region or any of the additional functional regions (TFBS, DNase, UCNE, dbSuper)</p>
</dd>
</dl>
</li>
</ul>
<p><strong>NB.</strong> Annotations requires GREEN-DB and accessory BED files to be present in the resources folder.
The default location is <code class="docutils literal notranslate"><span class="pre">resources/bed_files</span></code> within the tool folder, but you can provide a different location using <code class="docutils literal notranslate"><span class="pre">--bed_dir</span></code></p>
</div>
<div class="section" id="prioritization">
<h2>Prioritization<a class="headerlink" href="#prioritization" title="Permalink to this headline">¶</a></h2>
<p>If prioritization is active (-p, –prioritize), an additional NC_VARCLASS field will be added.
This fields is an integer from 0 to 13 wich summarize evidences supporting a regulatory impact for the variant.
Higher values are associated to a higher probability of regulatory impact.</p>
<p><strong>NB.</strong> You need ReMM, LinSight and NCBoost scores available to run priotization mode.
These scores will be automatically addedd to annotations when prioritization mode is activated.</p>
<p>Levels are based on GREEN-DB annotation and 3 prediction scores (ReMM, LinSight, NCBoost).
The FDR50 and TPR90 thresholds are defined in the GREEN-VARAN paper.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 8%" />
<col style="width: 25%" />
<col style="width: 23%" />
<col style="width: 18%" />
<col style="width: 26%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Level</p></th>
<th class="head"><p>1 score above threshold
(ReMM, LinSight, NCBoost)</p></th>
<th class="head"><p>Overlap GREEN-DB region</p></th>
<th class="head"><p>GREEN-DB
constraint &gt;= 0.9</p></th>
<th class="head"><p>Overlap any of
TFBS, DNase, UCNE, dbSuper</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>13</p></td>
<td><p>FDR50</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
</tr>
<tr class="row-odd"><td><p>12</p></td>
<td><p>FDR50</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>11</p></td>
<td><p>FDR50</p></td>
<td><p>X</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>10</p></td>
<td><p>TPR90</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
</tr>
<tr class="row-even"><td><p>9</p></td>
<td><p>TPR90</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>8</p></td>
<td><p>TPR90</p></td>
<td><p>X</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>7</p></td>
<td><p>FDR50</p></td>
<td></td>
<td></td>
<td><p>X</p></td>
</tr>
<tr class="row-odd"><td><p>6</p></td>
<td><p>TPR90</p></td>
<td></td>
<td></td>
<td><p>X</p></td>
</tr>
<tr class="row-even"><td><p>5</p></td>
<td></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
</tr>
<tr class="row-odd"><td><p>4</p></td>
<td></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td></td>
<td><p>X</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>FDR50</p></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>1</p></td>
<td><p>TPR90</p></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>0</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="gene-mode">
<h2>Gene mode<a class="headerlink" href="#gene-mode" title="Permalink to this headline">¶</a></h2>
<p>The gene mode allows to annotate / filter regulatory variants if they potentially affect a gene of interest.
Using the information on controlled / closest genes from GREEN-DB the tool will select only variants affecting a region
that is associated to gene(s) of interest.</p>
<p><strong>1. Activate the gene mode</strong></p>
<p>Gene mode is activated using the <code class="docutils literal notranslate"><span class="pre">-g,</span> <span class="pre">--gene_mode</span></code> option which can be set to:</p>
<ul class="simple">
<li><dl class="simple">
<dt>annotate</dt><dd><p>NC_VOI=1 annotation will be added to variants potentially affecting the gene(s) of interest</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>filter</dt><dd><p>Only variants potentially affecting the gene(s) of interest will be present in the output</p>
</dd>
</dl>
</li>
</ul>
<p>Using <code class="docutils literal notranslate"><span class="pre">-t,</span> <span class="pre">--gene_type</span></code> user can select which region-genes association should be considered to determine the affected genes.
Accepted values are:</p>
<ul class="simple">
<li><dl class="simple">
<dt>controlled</dt><dd><p>For each region, only gene(s) controlled according to experimental data in GREEN-DB will be considered</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>closest</dt><dd><p>For each region, only the closest gene(s) will be considered</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>both</dt><dd><p>For each region, both controlled and closest gene(s) are considered</p>
</dd>
</dl>
</li>
</ul>
<p><strong>2. Set gene of interest or impact</strong></p>
<p>When gene mode is active you can provide a list of genes of interest using <code class="docutils literal notranslate"><span class="pre">-l,</span> <span class="pre">--gene_list</span></code>.</p>
<p>The argument accepts a comma-separated list of gene symbols (like CFTR,BRCA1,BRCA2) or a text file containing genes one per line.
Regulatory variants associated to one of the gene in your list will be annotated / filtered as “variants of interest”</p>
<p>Using the <code class="docutils literal notranslate"><span class="pre">--impact</span></code> option, you can annotate / filter variants with a potential effect on a gene already affected
by a coding variants with a minimum impact.
The option accept the minimum impact level according to SnpEFF: HIGH,MODERATE,LOW,MODIFIER.
Note that when this option is active the tool will first scan your input VCF and collect the list of genes with at least 1 variant of the
given impact. This can slow down the whole process, since VCF need to be read twice.</p>
<p><strong>NB.</strong> Gene list and impact settings act together so if both are activated only variants passing both
will be considered as “variant of interest” and annotated / filtered</p>
</div>
<div class="section" id="activate-additional-annotations">
<h2>Activate additional annotations<a class="headerlink" href="#activate-additional-annotations" title="Permalink to this headline">¶</a></h2>
<div class="section" id="predictions-scores">
<h3>Predictions scores<a class="headerlink" href="#predictions-scores" title="Permalink to this headline">¶</a></h3>
<p>You can annotate variants using 7 different prediction scores. To activate one ore more of these annotations use the
<code class="docutils literal notranslate"><span class="pre">-s,</span> <span class="pre">--scores</span></code> option providing a single score name. The option can be repeated multiple time to add more scores.
Alternatively, you can set <code class="docutils literal notranslate"><span class="pre">--allscores</span></code> to activate all scores. Note that annotating with all scores can slow down the annotation
considerably. Our suggestion for rare variants is to use ReMM, NCBoost and LinSight.</p>
<p><strong>NB.</strong> Score annotations requires the corresponding tables from GREEN-VARAN release to be present in the resources folder.
Default location is <code class="docutils literal notranslate"><span class="pre">resources\scores</span></code> within the tool folder, but you can set a different one using <code class="docutils literal notranslate"><span class="pre">--scores_dir</span></code></p>
<p>Available scores included with the GREEN-VARAN release</p>
<ul class="simple">
<li><dl class="simple">
<dt>CADD v1.5</dt><dd><p><a class="reference external" href="https://academic.oup.com/nar/article/47/D1/D886/5146191">CADD: predicting the deleteriousness of variants throughout the human genome</a></p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>DANN</dt><dd><p><a class="reference external" href="https://academic.oup.com/bioinformatics/article/31/5/761/2748191">DANN: a deep learning approach for annotating the pathogenicity of genetic variants</a></p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>ExPECTO</dt><dd><p><a class="reference external" href="https://www.nature.com/articles/s41588-018-0160-6">Deep learning sequence-based ab initio prediction of variant effects on expression and disease risk</a></p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>FIRE</dt><dd><p><a class="reference external" href="https://academic.oup.com/bioinformatics/article/33/24/3895/4093216">FIRE: functional inference of genetic variants that regulate gene expression</a></p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>LinSight</dt><dd><p><a class="reference external" href="https://www.nature.com/articles/ng.3810">Fast, scalable prediction of deleterious noncoding variants from functional and population genomic data</a></p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>NCBoost</dt><dd><p><a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1634-2">NCBoost classifies pathogenic non-coding variants in Mendelian diseases through supervised learning on purifying selection signals in humans</a></p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>ReMM v0.3.1</dt><dd><p><a class="reference external" href="https://www.sciencedirect.com/science/article/pii/S0002929716302786">A Whole-Genome Analysis Framework for Effective Identification of Pathogenic Regulatory Variants in Mendelian Disease</a></p>
</dd>
</dl>
</li>
</ul>
</div>
<div class="section" id="population-allele-frequency">
<h3>Population allele frequency<a class="headerlink" href="#population-allele-frequency" title="Permalink to this headline">¶</a></h3>
<p>You can annotate population allele frequency from gnomAD genomes using <code class="docutils literal notranslate"><span class="pre">--allelefreq</span></code> to set the desired population.
The option accept standard population codes (afr,amr,eas,fin,nfe,sas,oth) or global for global AF.</p>
<p><strong>NB.</strong> This option requires gnomAD VCF file. A simplified version is provided with GREEN-VARAN release or you can specify
a different location using <code class="docutils literal notranslate"><span class="pre">--AF_file</span></code></p>
</div>
</div>
<div class="section" id="fields-added-to-info">
<h2>Fields added to INFO<a class="headerlink" href="#fields-added-to-info" title="Permalink to this headline">¶</a></h2>
<div class="section" id="green-db-related-fields">
<h3>GREEN-DB related fields<a class="headerlink" href="#green-db-related-fields" title="Permalink to this headline">¶</a></h3>
<p>Fields in the following table are added to INFO fields when <code class="docutils literal notranslate"><span class="pre">--separate_fields</span></code> option is active.
Otherwise, they are collpsed in a single NC_ANNO field, separated by pipe symbol <code class="docutils literal notranslate"><span class="pre">NC_ANNO=NC_support|NC_regionID|...</span></code></p>
<table class="colwidths-given docutils align-default">
<colgroup>
<col style="width: 20%" />
<col style="width: 20%" />
<col style="width: 60%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Annotation tag</p></th>
<th class="head"><p>Data type</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>NC_support</p></td>
<td><p>Float</p></td>
<td><p>Sum of max NC_constraint; NC_methods; NC_median_PhyloP100 positive values and binary values for presence/absence of NC_genes; NC_TFname; NC_DNase; NC_UCNE; NC_dbSUPER</p></td>
</tr>
<tr class="row-odd"><td><p>NC_regionID</p></td>
<td><p>String</p></td>
<td><p>Comma separated list of GREEN-DB region IDs for regions overlapping the variants</p></td>
</tr>
<tr class="row-even"><td><p>NC_region_type</p></td>
<td><p>String</p></td>
<td><p>Comma separated list of region types for regions overlapping the variants</p></td>
</tr>
<tr class="row-odd"><td><p>NC_constraint</p></td>
<td><p>Float</p></td>
<td><p>The maximum constraint value for GREEN-DB regions overlapping the variant</p></td>
</tr>
<tr class="row-even"><td><p>NC_methods</p></td>
<td><p>Integer</p></td>
<td><p>Number of methods supporting this location as regulatory regions; calculated as the number of distinct methods supporting the GREEN-DB regions overlapping the variant</p></td>
</tr>
<tr class="row-odd"><td><p>NC_genes</p></td>
<td><p>String</p></td>
<td><p>Comma separated list of controlled genes from GREEN-DB</p></td>
</tr>
<tr class="row-even"><td><p>NC_closestGene</p></td>
<td><p>String</p></td>
<td><p>Comma separated list of the closest genes from GREEN-DB</p></td>
</tr>
<tr class="row-odd"><td><p>NC_closestGene_dist</p></td>
<td><p>Integer</p></td>
<td><p>Comma separated list of the distance of the closest genes from GREEN-DB</p></td>
</tr>
<tr class="row-even"><td><p>NC_closestProt</p></td>
<td><p>String</p></td>
<td><p>Comma separated list of the closest protein-coding genes from GREEN-DB</p></td>
</tr>
<tr class="row-odd"><td><p>NC_closestProt_dist</p></td>
<td><p>Integer</p></td>
<td><p>Comma separated list of the distance of the closest protein-coding genes from GREEN-DB</p></td>
</tr>
<tr class="row-even"><td><p>NC_tolerant_P</p></td>
<td><p>Float</p></td>
<td><p>Maximum value of variant tolerant P across regions overlapping the variant</p></td>
</tr>
<tr class="row-odd"><td><p>NC_tolerant_label</p></td>
<td><p>String</p></td>
<td><p>Comma separated list of TOLERANT/INTOLERANT labels based on LoF tolerance probability across regions overlapping the variant</p></td>
</tr>
<tr class="row-even"><td><p>NC_median_PhyloP100</p></td>
<td><p>Float</p></td>
<td><p>Maximum value of median PhyloP100 across GREEN-DB regions overlapping the variant</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="additional-fields">
<h3>Additional fields<a class="headerlink" href="#additional-fields" title="Permalink to this headline">¶</a></h3>
<p>The following fields are always added as separated fields in the INFO column</p>
<table class="colwidths-given docutils align-default">
<colgroup>
<col style="width: 20%" />
<col style="width: 20%" />
<col style="width: 60%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Annotation tag</p></th>
<th class="head"><p>Data type</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>NC_TFname</p></td>
<td><p>String</p></td>
<td><p>Comma separated list of transcription factors binding at the variant location</p></td>
</tr>
<tr class="row-odd"><td><p>NC_DNase</p></td>
<td><p>Integer</p></td>
<td><p>Binary value representing the presence of a DNase HS site at the variant location</p></td>
</tr>
<tr class="row-even"><td><p>NC_UCNE</p></td>
<td><p>Integer</p></td>
<td><p>Binary value representing the presence of a UCNE at the variant location</p></td>
</tr>
<tr class="row-odd"><td><p>NC_dbSUPER</p></td>
<td><p>Integer</p></td>
<td><p>Binary value representing the presence of a dbSuper cluster at the variant location</p></td>
</tr>
<tr class="row-even"><td><p>NC_VOI</p></td>
<td><p>Integer</p></td>
<td><p>When gene mode is active this is set to one for variants overlapping a GREEN-DB region controlling a gene of interest</p></td>
</tr>
<tr class="row-odd"><td><p>NC_VARCLASS</p></td>
<td><p>Integer</p></td>
<td><p>When prioritize mode is active this value is set to the prioritization level (0-13)</p></td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="arguments-list">
<h2>Arguments list<a class="headerlink" href="#arguments-list" title="Permalink to this headline">¶</a></h2>
<div class="section" id="mandatory-arguments">
<h3>Mandatory Arguments<a class="headerlink" href="#mandatory-arguments" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd><span class="option">-h</span>, <span class="option">--help</span></kbd></dt>
<dd><div class="line-block">
<div class="line">Shows help message and exit</div>
</div>
</dd>
<dt><kbd><span class="option">-i <var>VCF</var></span>, <span class="option">--vcf <var>VCF</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">Input vcf[.gz] file</div>
</div>
</dd>
<dt><kbd><span class="option">-o <var>OUTPUT</var></span>, <span class="option">--output <var>OUTPUT</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">VCF output file (at the moment only support plain VCF output)</div>
</div>
</dd>
<dt><kbd><span class="option">-b <var>BUILD</var></span>, <span class="option">--build <var>BUILD</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">Possible values: <code class="docutils literal notranslate"><span class="pre">{GRCh37,GRCh38}</span></code></div>
<div class="line">Specify the genome build of input VCF</div>
</div>
</dd>
<dt><kbd><span class="option">-m <var>MODE</var></span>, <span class="option">--mode <var>MODE</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">Possible values: <code class="docutils literal notranslate"><span class="pre">{annotate,filter_regdb,filter_any}</span></code></div>
<div class="line">Set the running mode</div>
</div>
</dd>
</dl>
</div>
<div class="section" id="additional-annotations-scores-af">
<h3>Additional annotations (scores, AF)<a class="headerlink" href="#additional-annotations-scores-af" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd><span class="option">--allelefreq <var>POP_CODE</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">Possible values: <code class="docutils literal notranslate"><span class="pre">{global,afr,amr,eas,fin,nfe,sas,oth}</span></code></div>
<div class="line">Add gnomAD AF annotations based on global AF or specific population</div>
</div>
</dd>
<dt><kbd><span class="option">-s <var>SCORE_NAME</var></span>, <span class="option">--scores <var>SCORE_NAME</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">Possible values: <code class="docutils literal notranslate"><span class="pre">{ReMM,FIRE,LinSight,ExPECTO,NCBoost,DANN,CADD}</span></code></div>
<div class="line">Add selected prediction score for non-coding vars. Repeat to add multiple scores</div>
</div>
</dd>
<dt><kbd><span class="option">-a</span>, <span class="option">--allscores</span></kbd></dt>
<dd><div class="line-block">
<div class="line">Add all prediction score for non-coding vars (ReMM,FIRE,LinSight,ExPECTO,NCBoost)</div>
</div>
</dd>
</dl>
</div>
<div class="section" id="prioritize">
<h3>Prioritize<a class="headerlink" href="#prioritize" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd><span class="option">-p</span>, <span class="option">--prioritize</span></kbd></dt>
<dd><div class="line-block">
<div class="line">Turn on prioritization for non-coding variants</div>
</div>
</dd>
</dl>
</div>
<div class="section" id="gene-based-annotations">
<h3>Gene based annotations<a class="headerlink" href="#gene-based-annotations" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd><span class="option">-g <var>GENE_MODE</var></span>, <span class="option">--gene_mode <var>GENE_MODE</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">Possible values: <code class="docutils literal notranslate"><span class="pre">{annotate,filter}</span></code></div>
<div class="line">Activate gene based annotation/filter</div>
</div>
</dd>
<dt><kbd><span class="option">-t <var>GENE_TYPE</var></span>, <span class="option">--gene_type <var>GENE_TYPE</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">Possible values: {controlled,closest,both}</div>
<div class="line">DEFAULT: <code class="docutils literal notranslate"><span class="pre">controlled</span></code></div>
<div class="line">Which genes to consider for NC regions</div>
</div>
</dd>
<dt><kbd><span class="option">-l <var>GENES_LIST</var></span>, <span class="option">--genes_list <var>GENES_LIST</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">List of genes of interest, can be comma-separated list or file with one gene per line</div>
</div>
</dd>
<dt><kbd><span class="option">--impact <var>MIN_IMPACT</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">Possible values: <code class="docutils literal notranslate"><span class="pre">{HIGH,MODERATE,LOW,MODIFIER}</span></code></div>
<div class="line">Only report NC vars if the controlled at least this impact</div>
</div>
</dd>
</dl>
</div>
<div class="section" id="customize-files-locations">
<h3>Customize files locations<a class="headerlink" href="#customize-files-locations" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd><span class="option">--vcfanno <var>VCFANNO</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">Full path to vcfanno executable</div>
</div>
</dd>
<dt><kbd><span class="option">--bed_dir <var>BED_DIR</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">Directory containing RegDB bed files</div>
</div>
</dd>
<dt><kbd><span class="option">--AF_file <var>AF_FILE</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">Full path to gnomAD VCF for AF annotation</div>
</div>
</dd>
<dt><kbd><span class="option">--scores_dir <var>SCORES_DIR</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">Directory containing prediction scores tables</div>
</div>
</dd>
<dt><kbd><span class="option">--logfile <var>LOGFILE</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">Log file</div>
</div>
</dd>
</dl>
</div>
<div class="section" id="additional-arguments">
<h3>Additional Arguments<a class="headerlink" href="#additional-arguments" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd><span class="option">--separate_fields</span></kbd></dt>
<dd><div class="line-block">
<div class="line">Make multiple fields instead of a single NC_ANNO annotation</div>
</div>
</dd>
<dt><kbd><span class="option">--threads <var>THREADS</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">Number of threads for vcfanno annotation</div>
</div>
</dd>
<dt><kbd><span class="option">-w</span>, <span class="option">--overwrite</span></kbd></dt>
<dd><div class="line-block">
<div class="line">If set, overwrite output file if already exists</div>
</div>
</dd>
</dl>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="SV_annotation_usage.html" class="btn btn-neutral float-right" title="SV_annotation tool usage" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="GREEN_VARAN.html" class="btn btn-neutral float-left" title="GREEN-VARAN tool set" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Edoardo Giacopuzzi

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>