

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>GREEN-VARAN tool usage &mdash; GREEN-DB and GREEN-VARAN  documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="greendb_query tool usage" href="GREEN-DB_query_usage.html" />
    <link rel="prev" title="GREEN-VARAN tool set" href="GREEN_VARAN.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> GREEN-DB and GREEN-VARAN
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="GREEN_DB.html">The GREEN-DB</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="GREEN_VARAN.html">GREEN-VARAN tool set</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="GREEN_VARAN.html#installation">Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="GREEN_VARAN.html#get-green-db-files">Get GREEN-DB files</a></li>
<li class="toctree-l2"><a class="reference internal" href="GREEN_VARAN.html#green-varan-nextflow-workflow">GREEN-VARAN Nextflow workflow</a></li>
<li class="toctree-l2"><a class="reference internal" href="GREEN_VARAN.html#singularity">Singularity</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="GREEN_VARAN.html#single-tools-usage">Single tools usage</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">GREEN-VARAN tool usage</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#basic-usage">Basic usage</a></li>
<li class="toctree-l4"><a class="reference internal" href="#command-line-options">Command line options</a></li>
<li class="toctree-l4"><a class="reference internal" href="#annotations-added-by-green-varan">Annotations added by GREEN-VARAN</a></li>
<li class="toctree-l4"><a class="reference internal" href="#prioritization-of-small-variants">Prioritization of small variants</a></li>
<li class="toctree-l4"><a class="reference internal" href="#structural-variants-annotations">structural variants annotations</a></li>
<li class="toctree-l4"><a class="reference internal" href="#singularity">Singularity</a></li>
<li class="toctree-l4"><a class="reference internal" href="#example-usage">Example usage</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="GREEN-DB_query_usage.html">greendb_query tool usage</a></li>
<li class="toctree-l3"><a class="reference internal" href="nextflow_workflow.html">GREEN-VARAN workflow</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="nextflow_workflow.html">GREEN-VARAN workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="Download.html">Download resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="how_to_cite.html">How to cite</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">GREEN-DB and GREEN-VARAN</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="GREEN_VARAN.html">GREEN-VARAN tool set</a> &raquo;</li>
        
      <li>GREEN-VARAN tool usage</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/GREEN-VARAN_usage.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="green-varan-tool-usage">
<h1>GREEN-VARAN tool usage<a class="headerlink" href="#green-varan-tool-usage" title="Permalink to this headline">¶</a></h1>
<p>GREEN-VARAN performs annotation of small variants or structural variants VCF adding information on potential regulatory variants from GREEN-DB.
Especially it can annotate possible controlled genes and a prioritization level (this latter need the presence of some additional annotations, see below)
It provides also abiliy to tag variants linked to genes of interest and update existing gene-level annotations from SnpEff or bcftools.</p>
<div class="section" id="basic-usage">
<h2>Basic usage<a class="headerlink" href="#basic-usage" title="Permalink to this headline">¶</a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>greenvaran <span class="o">[</span>run mode<span class="o">]</span> <span class="o">[</span>options<span class="o">]</span>
</pre></div>
</div>
<p>The running mode can be one of:</p>
<ul class="simple">
<li><p>smallvars
In this mode the tool will perform annotation for a small variants VCF.
It will annotate variants with information on the possible regulatory role based on GREENDB and eventually provide prioritization levels</p></li>
<li><p>sv
In this mode the tool will perform annotation for a structural variants VCF.
Capability in this case is limited to annotation of overlapping GREENDB regions and controlled genes. No prioritization is provided</p></li>
<li><p>querytab
This mode is a convenient way to automatically prepare input table to be used with the query tool to exctract detailed information from GREENDB database.</p></li>
<li><p>version
Print the tool version</p></li>
</ul>
<p><strong>NB.</strong> To perform prioritization of small variants some additional annotation fields are expected in the input VCF, see the prioritization section below.
By default, when these information are not present the prioritization level will be set to zero for all annotated variants.
We also provide pre-processed datasets and Nextflow workflow to automate the whole process (see #TODO nextflow workflow page).</p>
</div>
<div class="section" id="command-line-options">
<h2>Command line options<a class="headerlink" href="#command-line-options" title="Permalink to this headline">¶</a></h2>
<div class="section" id="smallvars-and-sv-shared-options">
<h3>smallvars and sv shared options<a class="headerlink" href="#smallvars-and-sv-shared-options" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd><span class="option">-i</span>, <span class="option">--invcf <var>INVCF</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">path to indexed input vcf.gz/bcf.</div>
</div>
</dd>
<dt><kbd><span class="option">-o</span>, <span class="option">--outvcf <var>OUTVCF</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">output vcf / vcf.gz file</div>
</div>
</dd>
<dt><kbd><span class="option">-d</span>, <span class="option">--db <var>DB</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">GREEN-DB bed.gz file for your build (see download section)</div>
</div>
</dd>
<dt><kbd><span class="option">-s</span>, <span class="option">--dbschema <var>DBSCHEMA</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">json file containig greendb column mapping</div>
<div class="line">A default configuration for GREENDB v2.5 is available in config folder</div>
</div>
</dd>
<dt><kbd><span class="option">-u</span>, <span class="option">--noupdate</span></kbd></dt>
<dd><div class="line-block">
<div class="line">do not update ANN / BCSQ field in the input VCF</div>
</div>
</dd>
<dt><kbd><span class="option">-f</span>, <span class="option">--filter</span></kbd></dt>
<dd><div class="line-block">
<div class="line">filter instead of annotate. Only variants with greendb overlap will be written.</div>
<div class="line">If –genes is active, the output will contain only variants connected to the input genes of interest</div>
</div>
</dd>
<dt><kbd><span class="option">-m</span>, <span class="option">--impact <var>IMPACT</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">Which impact to assign when updating snpEff field</div>
<div class="line">Possible values: [HIGH, MODERATE, LOW, MODIFIER] (default: MODIFIER)</div>
</div>
</dd>
<dt><kbd><span class="option">--chrom <var>CHROM</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">Annotate only for a specific chromosome</div>
<div class="line">Useful to parallelize across chromosomes</div>
</div>
</dd>
<dt><kbd><span class="option">-g</span>, <span class="option">--genes <var>GENES</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">Gene symbols for genes of interest, variants connected to those will be flagged with greendb_VOI tag</div>
<div class="line">This can be a comma-separated list or a text file listing genes one per line</div>
</div>
</dd>
<dt><kbd><span class="option">--connection <var>CONNECTION</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">Region-gene connections accepted for annotation</div>
<div class="line">Possible values: [all, closest, annotated] (default: all)</div>
</div>
</dd>
<dt><kbd><span class="option">--log <var>LOG</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">Log file. Default is greenvaran_[now].log</div>
</div>
</dd>
</dl>
</div>
<div class="section" id="sv-specific-options">
<h3>sv specific options<a class="headerlink" href="#sv-specific-options" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd><span class="option">-p</span>, <span class="option">--padding <var>PADDING</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">Value to add on each side of BND/INS, this override the CIPOS when set</div>
</div>
</dd>
<dt><kbd><span class="option">--cipos <var>CIPOS</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">INFO field listing the confidence interval around breakpoints</div>
<div class="line">It is expected to have 2 comma-separated values (default: CIPOS)</div>
</div>
</dd>
<dt><kbd><span class="option">-t</span>, <span class="option">--minoverlap <var>MINOVERLAP</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">Min fraction of GREENDB region to be overlapped by a SV (default: 0.000001)</div>
</div>
</dd>
<dt><kbd><span class="option">-b</span>, <span class="option">--minbp <var>MINBP</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">Min number of bases of GREENDB region to be overlapped by a SV (default: 1)</div>
</div>
</dd>
</dl>
</div>
<div class="section" id="smallvars-specific-options">
<h3>smallvars specific options<a class="headerlink" href="#smallvars-specific-options" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd><span class="option">-c</span>, <span class="option">--config <var>CONFIG</var></span></kbd></dt>
<dd><div class="line-block">
<div class="line">json config file for prioritization</div>
<div class="line">A default configuration for the four level described in the paper is provided in config folder</div>
</div>
</dd>
<dt><kbd><span class="option">-p</span>, <span class="option">--permissive</span></kbd></dt>
<dd><div class="line-block">
<div class="line">Perform prioritization even if one of the INFO fields required by prioritization config is missing</div>
<div class="line">By default, when one of the expeced fields is not defined in the header, the prioritization is disabled and all variants will get level zero</div>
</div>
</dd>
</dl>
</div>
</div>
<div class="section" id="annotations-added-by-green-varan">
<h2>Annotations added by GREEN-VARAN<a class="headerlink" href="#annotations-added-by-green-varan" title="Permalink to this headline">¶</a></h2>
<div class="section" id="info-fields">
<h3>INFO fields<a class="headerlink" href="#info-fields" title="Permalink to this headline">¶</a></h3>
<p>Fields in the following table are added to INFO fields by GREEN-VARAN. greendb_level will be added only for small variants</p>
<table class="colwidths-given docutils align-default">
<colgroup>
<col style="width: 20%" />
<col style="width: 20%" />
<col style="width: 60%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Annotation tag</p></th>
<th class="head"><p>Data type</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>greendb_id</p></td>
<td><p>String</p></td>
<td><p>Comma-separated list of GREEN-DB IDs identifying the regions that overlap this variant</p></td>
</tr>
<tr class="row-odd"><td><p>greendb_stdtype</p></td>
<td><p>String</p></td>
<td><p>Comma-separated list of standard region types as annotated in GREEN-DB for regions overlapping the variant</p></td>
</tr>
<tr class="row-even"><td><p>greendb_dbsource</p></td>
<td><p>String</p></td>
<td><p>Comma-separated list of data sources as annotated in GREEN-DB for regions overlapping the variant</p></td>
</tr>
<tr class="row-odd"><td><p>greendb_level</p></td>
<td><p>Integer</p></td>
<td><p>Variant prioritization level computed by GREEN-VARAN. See Prioritization section below</p></td>
</tr>
<tr class="row-even"><td><p>greendb_constraint</p></td>
<td><p>Float</p></td>
<td><p>The maximum constraint value across GREEN-DB regions overlapping the variant</p></td>
</tr>
<tr class="row-odd"><td><p>greendb_genes</p></td>
<td><p>String</p></td>
<td><p>Possibly controlled genes for regulatory regions overlapping this variant</p></td>
</tr>
<tr class="row-even"><td><p>greendb_VOI</p></td>
<td><p>Flag</p></td>
<td><p>When <code class="docutils literal notranslate"><span class="pre">--genes</span></code> option is active this flag is set when any of the input genes is among the possibly controlled genes for overlapping regulatory regions.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="updated-gene-consequences">
<h3>Updated gene consequences<a class="headerlink" href="#updated-gene-consequences" title="Permalink to this headline">¶</a></h3>
<p>By default, GREEN-VARAN update gene consequences in the SnpEff ANN field or the bcftools BCSQ if one is present in the input VCF file.
In this way the annotation can be processed by most downstream tools evaluating segregation.
If none is found, GREEN-VARAN will create a new ANN field. To switch off gene consequence update use the <code class="docutils literal notranslate"><span class="pre">--noupdate</span></code> option.</p>
<p>Here the tool will add one a new consequence for each possibly controlled genes, limited by the <code class="docutils literal notranslate"><span class="pre">--connection</span></code> option.
The new consequence will follow standard format according to SnpEff or bcftools and have MODIFIER impact by default.
This can be adjusted using the <code class="docutils literal notranslate"><span class="pre">--impact</span></code> option.
The gene effect will be set according to the GREEN-DB region type, adding 5 new terms: <cite>bivalent, enhancer, insulator, promoter, silencer</cite>.</p>
<p>Example ANN / BCSQ field added by GREEN-VARAN.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nv">ANN</span><span class="o">=</span>C<span class="p">|</span>enhancer<span class="p">|</span>MODIFIER<span class="p">|</span>GeneA<span class="o">||||||||||||</span>
<span class="nv">BCQS</span><span class="o">=</span>enhancer<span class="p">|</span>GeneA<span class="o">||</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="prioritization-of-small-variants">
<h2>Prioritization of small variants<a class="headerlink" href="#prioritization-of-small-variants" title="Permalink to this headline">¶</a></h2>
<p>GREEN-VARAN will consider GREEN-DB annotations, additional functional regions and non-coding impact prediction scores to provide a prioritization level for each annotated variant.
This level is annotated under greenvara_level tag in the INFO field.
This fields is an integer from 0 to N wich summarize evidences supporting a regulatory impact for the variant.
Higher values are associated to a higher probability of regulatory impact.</p>
<p><strong>NB.</strong> You need teh following INFO fields in your input VCF to run priotization mode as described in the GREEN-DB manuscript
using the default config provided.</p>
<ol class="arabic simple">
<li><p>gnomAD_AF, gnomAD_AF_nfe float values describing global and NFE population AF from gnomAD</p></li>
<li><p>ncER, FATHMM-MKL and ReMM float values providing scores predictions</p></li>
<li><p>TFBS, DNase and UCNE flags describing overlap with additional functional regions</p></li>
</ol>
<p>This configuration resembles the four levels prioritization described in the GREEN-DB manuscript.
Note that the exact names of these annotations and the score thresholds are defined in the json file passed to –config options.</p>
<p>The following table summarizes the four prioritization levels defined in the manuscript and in the default config file.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 4%" />
<col style="width: 96%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Level</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>Rare variant (population AF &lt; 1%) overlapping one of GREEN-DB regions</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>Level 1 criteria and overlap at least one functional element among transcription factors binding sites (TFBS), DNase peaks, ultra conserved elements (UCNE)</p></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p>Level 2 criteria and prediction score value above the suggested FDR50 threshold for at least one among ncER, FATHMM MKL, ReMM</p></td>
</tr>
<tr class="row-odd"><td><p>4</p></td>
<td><p>Level 3 critera and region constraint value greater or equal 0.7</p></td>
</tr>
</tbody>
</table>
<div class="section" id="personalize-the-prioritization-schema">
<h3>Personalize the prioritization schema<a class="headerlink" href="#personalize-the-prioritization-schema" title="Permalink to this headline">¶</a></h3>
<p>The prioritization schema is defined in a config json file. The default is provided in the config folder.
An example of expected file structure is reported below</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="o">{</span>
    <span class="s2">&quot;af&quot;</span>: <span class="o">[</span><span class="s2">&quot;gnomAD_AF&quot;</span>,<span class="s2">&quot;gnomAD_AF_nfe&quot;</span><span class="o">]</span>,
    <span class="s2">&quot;maxaf&quot;</span>: <span class="m">0</span>.01,
    <span class="s2">&quot;regions&quot;</span>: <span class="o">[</span><span class="s2">&quot;TFBS&quot;</span>, <span class="s2">&quot;DNase&quot;</span>, <span class="s2">&quot;UCNE&quot;</span><span class="o">]</span>,
    <span class="s2">&quot;scores&quot;</span>: <span class="o">{</span>
        <span class="s2">&quot;FATHMM_MKLNC&quot;</span>: <span class="m">0</span>.908,
        <span class="s2">&quot;ncER&quot;</span>: <span class="m">98</span>.6,
        <span class="s2">&quot;ReMM&quot;</span>: <span class="m">0</span>.963
    <span class="o">}</span>,
    <span class="s2">&quot;constraint&quot;</span>: <span class="m">0</span>.7,
    <span class="s2">&quot;more_regions&quot;</span>: <span class="o">[]</span>,
    <span class="s2">&quot;more_values&quot;</span>: <span class="o">{}</span>
<span class="o">}</span>
</pre></div>
</div>
<p>Sections definitions:</p>
<ol class="arabic simple">
<li><p>af: INFO fields containing AF annotations. The tool will consider the max value across all these</p></li>
<li><p>maxaf: if the max value across af fields is below this, the variant get +1 point</p></li>
<li><p>regions: INFO fields for overlapping regions. If any of these is set, the variant get +1 point</p></li>
<li><p>scores: series of key, value pairs. If any of key value is above the configured value, the variant get +1 point</p></li>
<li><p>constraint: if the max constraint value across overlapping GREEN-DB regions is above this value, the variant get +1 point</p></li>
<li><p>more_regions: any additional INFO fields representing overlap with custom regions. The variant get +1 point for each positive overlap</p></li>
<li><p>more_values: series of key, value pairs. The variant get +1 point fro each key value above the configured value</p></li>
</ol>
<p><strong>NB.</strong> more_regions and more_values must always been present. Leave them empty like in the example above if you don’t want to configure any custom value.</p>
<p><strong>NB2.</strong> INFO fields specified by af, scores and more_values are expected to be float, while those specified by regions and more_regions are expected as flags.</p>
</div>
</div>
<div class="section" id="structural-variants-annotations">
<h2>structural variants annotations<a class="headerlink" href="#structural-variants-annotations" title="Permalink to this headline">¶</a></h2>
<p>The annotation of structural variants is based on overlap with the regulatory regions defined in GREEN-DB.
This is treated differently according to the SV type:</p>
<ul class="simple">
<li><p>For <strong>DEL, DUP, INV</strong> an interval is constructed based on position field and the END info field from INFO.
When END is missing, the tool will try to use SVLEN instead. If none is not found the variant is not annotated
The user can then set a minimum level of overlap as either overlap fraction (<code class="docutils literal notranslate"><span class="pre">--minoverlap</span></code>) or N bp overlap (<code class="docutils literal notranslate"><span class="pre">--minbp</span></code>).
A GREEN-DB region is added to annotation only if its overlapping porting is larger or equal to both threshold</p></li>
<li><p>For <strong>INS and BND</strong>, an interval is constructed using the position and the coordinates in the CIPOS field (an alternative field can be set using <code class="docutils literal notranslate"><span class="pre">--cipos</span></code>).
This is done since INS and BND are often represented as single positions in structural variants VCF.
Alternatively, the user can provide a padding values using <code class="docutils literal notranslate"><span class="pre">--padding</span></code> and this value will be added aroud position
For these kind of variants any overlapping GREEN-DB region will be reported, diregarding the overlap threasholds</p></li>
</ul>
</div>
<div class="section" id="singularity">
<h2>Singularity<a class="headerlink" href="#singularity" title="Permalink to this headline">¶</a></h2>
<p>The tool binaries should work on most linux based system. In case you have any issue, we also provdie GREEN-VARAN as Singularity image (tested on singularity &gt;= 3.2).
A Singularity recipe is included in the repository or you can pull the image from Singularity Library using</p>
<p><code class="docutils literal notranslate"><span class="pre">singularity</span> <span class="pre">pull</span> <span class="pre">library://edg1983/greenvaran/greenvaran:latest</span></code></p>
<div class="section" id="usage">
<h3>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h3>
<p>The image contains both greenvaran and greendb_query tools.
The general usage is:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>singularity <span class="nb">exec</span> <span class="se">\</span>
greenvaran.sif <span class="se">\</span>
tool_name <span class="o">[</span>tool arguments<span class="o">]</span>
</pre></div>
</div>
</div>
<div class="section" id="bind-specific-folders-for-resources-or-data">
<h3>Bind specific folders for resources or data<a class="headerlink" href="#bind-specific-folders-for-resources-or-data" title="Permalink to this headline">¶</a></h3>
<p>The tool needs access to input VCF file, required GREEN-DB bed file and config files so remember to bind the corresponding locations in the container</p>
<p>See the following example where we use the current working directory for input/output, while other files are located
in the default config / resources folder within greenvaran folder. In the example we use GRCh38 genome build</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>singularity <span class="nb">exec</span> <span class="se">\</span>
--bind /greenvaran_path/resources/GRCh38:/db_files <span class="se">\</span>
--bind /greenvaran_path/config:/config_files <span class="se">\</span>
--bind <span class="si">${</span><span class="nv">PWD</span><span class="si">}</span>:/data <span class="se">\</span>
greenvaran.sif <span class="se">\</span>
greenvaran -i /data/input.vcf.gz <span class="se">\</span>
-o /data/output.vcf.gz <span class="se">\</span>
--db /db_files/GRCh38_GREEN-DB.bed.gz <span class="se">\</span>
--dbschema /config_files/greendb_schema_v2.5.json <span class="se">\</span>
--config /config_files/prioritize_smallvars.json
<span class="o">[</span>additional tool arguments<span class="o">]</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="example-usage">
<h2>Example usage<a class="headerlink" href="#example-usage" title="Permalink to this headline">¶</a></h2>
<div class="section" id="small-variants-test">
<h3>small variants test<a class="headerlink" href="#small-variants-test" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>greenvaran smallvars <span class="se">\</span>
--invcf test/VCF/GRCh38.test.smallvars.vcf.gz <span class="se">\</span>
--outvcf test/out/smallvars.annotated.vcf.gz <span class="se">\</span>
--config config/prioritize_smallvars.json <span class="se">\</span>
--dbschema config/greendb_schema_v2.5.json <span class="se">\</span>
--db resources/GRCh38/GRCh38_GREEN-DB.bed.gz <span class="se">\</span>
--genes test/VCF/genes_list_example.txt
</pre></div>
</div>
</div>
<div class="section" id="structural-variants-test">
<h3>structural variants test<a class="headerlink" href="#structural-variants-test" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>greenvaran sv <span class="se">\</span>
--invcf test/VCF/GRCh38.test.SV.vcf.gz <span class="se">\</span>
--outvcf test/out/SV.annotated.vcf.gz <span class="se">\</span>
--dbschema config/greendb_schema_v2.5.json <span class="se">\</span>
--db resources/GRCh38/GRCh38_GREEN-DB.bed.gz <span class="se">\</span>
--minbp <span class="m">10</span>
</pre></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="GREEN-DB_query_usage.html" class="btn btn-neutral float-right" title="greendb_query tool usage" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="GREEN_VARAN.html" class="btn btn-neutral float-left" title="GREEN-VARAN tool set" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Edoardo Giacopuzzi

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>