How to cite
===========

GREEN-DB
~~~~~~~~
If you use any information from GREEN-DB please cite:
`GREEN-DB: a framework for the annotation and prioritization of non-coding regulatory variants in whole-genome sequencing <https://www.biorxiv.org/content/10.1101/2020.09.17.301960v1>`_
Giacopuzzi E., Popitsch N., Taylor JC. BiorXiv (2020)

GREEN-VARAN
~~~~~~~~~~~
When you use GREEN-VARAN.py for annotation please cite

| 1. `GREEN-DB: a framework for the annotation and prioritization of non-coding regulatory variants in whole-genome sequencing <https://www.biorxiv.org/content/10.1101/2020.09.17.301960v1>`_
|    Giacopuzzi E., Popitsch N., Taylor JC. BiorXiv (2020)
| 2. `Vcfanno: fast, flexible annotation of genetic variants <https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0973-5>`_ 
|    Brent S. Pedersen, Ryan M. Layer & Aaron R. Quinlan. Genome Biology volume 17, Article number: 118 (2016)

If you include annotation with a prediction score please also cite the corresponding paper

| **CADD**: `CADD: predicting the deleteriousness of variants throughout the human genome <https://academic.oup.com/nar/article/47/D1/D886/5146191>`_
| **DANN**: `DANN: a deep learning approach for annotating the pathogenicity of genetic variants <https://academic.oup.com/bioinformatics/article/31/5/761/2748191>`_
| **ExPECTO**: `Deep learning sequence-based ab initio prediction of variant effects on expression and disease risk <https://www.nature.com/articles/s41588-018-0160-6>`_
| **FIRE**: `FIRE: functional inference of genetic variants that regulate gene expression <https://academic.oup.com/bioinformatics/article/33/24/3895/4093216>`_
| **LinSight**: `Fast, scalable prediction of deleterious noncoding variants from functional and population genomic data <https://www.nature.com/articles/ng.3810>`_
| **NCBoost**: `NCBoost classifies pathogenic non-coding variants in Mendelian diseases through supervised learning on purifying selection signals in humans <https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1634-2>`_
| **ReMM**: `A Whole-Genome Analysis Framework for Effective Identification of Pathogenic Regulatory Variants in Mendelian Disease <https://www.sciencedirect.com/science/article/pii/S0002929716302786>`_

Population AF
~~~~~~~~~~~~~
If you use SV_annotation or population AF annotation in GREEN-VARAN also cite the gnomAD paper:
`The mutational constraint spectrum quantified from variation in 141,456 humans <https://www.nature.com/articles/s41586-020-2308-7>`_